### PR TITLE
Feature autodoc stormtypes

### PR DESCRIPTION
**Pull request by vEpiphyte**
- Move storm prims and libs to use registration decorators to add them to a process global registry
- Convert addLibFuncs calls to generally using a getObjLocals method which just returns a dictionary of object locals for introspection purposes
- Add stormlib inspection and documentation generation to autodoc tool
- Add docstrings to all stormlib functions
- Add documentation output to include autodocs for stormlibs and stormprims

Not included:
- Updated docstrings for storm prims

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -235,6 +245,8 @@ def convert_ipynb(_):
             dirs.remove('.ipynb_checkpoints')
         for fn in fns:
             if fn.endswith('.ipynb'):
+                if 'httpapi' not in fn:
```
Comment (vEpiphyte) [inline docs/conf.py:248]: remove me

File: synapse/lib/stormhttp.py [hunk 2]
```diff
@@ -6,17 +6,20 @@
...
-        self.locls.update({
+    _storm_lib_path = ('inet', 'http')
+
+    def getObjLocals(self):
+        return {
             'get': self._httpEasyGet,
             'post': self._httpPost,
             # 'session':
-        })
+        }
```
Comment (vEpiphyte) [inline synapse/lib/stormhttp.py:22]: add docs

File: synapse/lib/stormlib/model.py [hunk 3]
```diff
@@ -133,12 +168,14 @@ def _getFormProp(self, name):
...
             return ModelProp(prop)
 
+@s_stormtypes.registry.registerType
 class ModelProp(s_stormtypes.Prim):
 
     def __init__(self, prop, path=None):
 
         s_stormtypes.Prim.__init__(self, prop, path=path)
 
+        # FIXME name / full
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/model.py:178]: resolve this

File: synapse/lib/stormlib/model.py [hunk 4]
```diff
@@ -163,9 +201,15 @@ def __init__(self, valu, path=None):
         s_stormtypes.Prim.__init__(self, valu, path=path)
         self.locls.update({
             'name': valu.name,
+        })
+        self.locls.update(self.getObjLocals())
+
+    # fixme name
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/model.py:207]: resolve this

File: synapse/tools/autodoc.py [hunk 5]
```diff
@@ -536,6 +538,135 @@ def get(self, key):
...
+                funcpath = '.'.join((libpath, name))
+                header = f'${funcpath}{callsig}'
+                header = header.replace('*', '\*')
+
+                libspage.addHead(header, lvl=2)
+
+                libspage.addLines(*newlines)
+
+            else:
+                # TODO - Handle package constants
```
Comment (vEpiphyte) [inline synapse/tools/autodoc.py:645]: resolve this

File: synapse/lib/stormtypes.py [hunk 6]
```diff
@@ -379,6 +647,19 @@ def _get_mesg(mesg, **kwargs):
         return mesg
 
     async def _print(self, mesg, **kwargs):
+        '''
+        Print a message to the runtime.
+
+        Args:
+            mesg (str): String to print.
+            **kwargs: Words go here FIXME
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:655]: resolve this

File: synapse/lib/stormtypes.py [hunk 7]
```diff
@@ -401,26 +682,84 @@ async def _pprint(self, item, prefix='', clamp=None):
                 await self.runt.printf(fline)
 
     async def _warn(self, mesg, **kwargs):
+        '''
+        Print a warning message to the runtime.
+
+        Args:
+            mesg (str): String to warn.
+            **kwargs: Words go here FIXME
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:690]: resolve this

File: synapse/lib/stormtypes.py [hunk 8]
```diff
@@ -1587,16 +2310,23 @@ async def __aiter__(self):
...
 class Path(Prim):
 
     def __init__(self, node, path=None):
         Prim.__init__(self, node, path=path)
+        self.locls.update(self.getObjLocals())
         self.locls.update({
+            'vars': PathVars(path),
+        })
+
+    # Fixme get vars out for docs purposes
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2323]: resolve this

File: synapse/lib/stormtypes.py [hunk 9]
```diff
@@ -1774,12 +2551,18 @@ def __init__(self, runt, ldef, path=None):
         self.runt = runt
         self.locls.update({
             'iden': ldef.get('iden'),
+        })
+        self.locls.update(self.getObjLocals())
+
+    # fixme plumb in iden
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2557]: resolve this

File: synapse/lib/stormtypes.py [hunk 10]
```diff
@@ -1878,10 +2692,15 @@ def __init__(self, runt, vdef, path=None):
...
         self.runt = runt
         self.locls.update({
-
             'iden': vdef.get('iden'),
             'layers': [Layer(runt, ldef, path=path) for ldef in vdef.get('layers')],
             'triggers': [Trigger(runt, tdef) for tdef in vdef.get('triggers')],
+        })
+        self.locls.update(self.getObjLocals())
+
+    # FIXME plumb in iden/layers/triggers
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2701]: resolve this

File: synapse/lib/stormtypes.py [hunk 11]
```diff
@@ -2183,113 +3070,236 @@ def ruleFromText(text):
...
 
+@registry.registerType
 class Gate(Prim):
 
     def __init__(self, runt, valu, path=None):
 
         Prim.__init__(self, valu, path=path)
         self.runt = runt
 
+        # FIXME get iden/roles/users
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:3295]: resolve this

File: synapse/lib/stormtypes.py [hunk 12]
```diff
@@ -2299,20 +3309,25 @@ def __init__(self, runt, valu, path=None):
 
         self.locls.update({
             'iden': valu,
+        })
+        self.locls.update(self.getObjLocals())
+
+    # Fixme iden
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:3315]: resolve this

File: synapse/lib/stormtypes.py [hunk 13]
```diff
@@ -2852,9 +3926,15 @@ def __init__(self, runt, cdef, path=None):
         self.runt = runt
         self.locls.update({
             'iden': cdef.get('iden'),
+        })
+        self.locls.update(self.getObjLocals())
+
+    # FIXME iden
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:3932]: resolve this

File: synapse/tools/autodoc.py [hunk 14]
```diff
@@ -536,6 +538,135 @@ def get(self, key):
...
+        print(obj)
+
+    # FIXME add links!
+    typespage = RstHelp
+    libspage = RstHelp()
+
+    libspage.addHead('Storm Libraries', lvl=0)
+
+    libspage.addLines('',
+                      'Storm Libraries are magical unicorns of the Storm query language.',
```
Comment (vEpiphyte) [inline synapse/tools/autodoc.py:572]: Do not use filler text.

File: docs/index.rst [hunk 15]
```diff
@@ -26,6 +26,8 @@ Welcome to the Synapse documentation!
    synapse/autodocs/datamodel_types
    synapse/autodocs/datamodel_forms
 
+   synapse/autodocs/stormtypes_libs
```
Comment (vEpiphyte) [inline docs/index.rst:29]: Identify a proper location for the libs

File: synapse/lib/stormlib/model.py [hunk 16]
```diff
@@ -76,40 +76,70 @@
...
-    def addLibFuncs(self):
-        self.locls.update({
+    def getObjLocals(self):
+        return {
             'type': self._methType,
             'prop': self._methProp,
             'form': self._methForm,
-            'edge': ModelEdge(self.runt),
-        })
+        }
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/model.py:95]: This was converted into a StormLib

File: synapse/lib/stormtypes.py [hunk 17]
```diff
@@ -268,20 +455,73 @@ def addLibFuncs(self):
...
             'cast': self._cast,
             'warn': self._warn,
             'print': self._print,
             'pprint': self._pprint,
             'sorted': self._sorted,
             'import': self._libBaseImport,
-        })
+        }
+
+    @property
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:470]: using @property allows us to get an inspectable Property object during documentation generation; but the value is properly resolved during runtime

File: synapse/lib/stormhttp.py [hunk 18]
```diff
@@ -35,6 +49,23 @@ async def _httpEasyGet(self, url, headers=None, ssl_verify=True):
                 return HttpResp(info)
 
     async def _httpPost(self, url, headers=None, json=None, body=None, ssl_verify=True):
+        '''
+        Post data to a given URL.
+
+        Args:
+            url (str): The URL to post too.
```
Comment (Cisphyx) [inline synapse/lib/stormhttp.py:56]: 'to post to'

File: synapse/lib/stormlib/macro.py [hunk 19]
```diff
@@ -124,27 +124,56 @@ async def wrapgenr():
...
+            dict: A macro definition.
+        '''
         name = await s_stormtypes.tostr(name)
 
         path = ('cortex', 'storm', 'macros', name)
         return await self.runt.snap.core.getHiveKey(path)
 
     async def _funcMacroDel(self, name):
+        '''
+        Delete a Storm Macros by name from the Cortex.
```
Comment (Cisphyx) [inline synapse/lib/stormlib/macro.py:169]: extra s on 'Macros'

File: synapse/lib/stormlib/model.py [hunk 20]
```diff
@@ -215,14 +261,45 @@ async def _chkKeyName(self, key):
             raise s_exc.NoSuchProp(mesg=f'The requested key is not valid for light edge metadata.',
                                    name=key)
 
+    def _methValidKeys(self):
+        '''
+        Get a list of the valid keys that can be set on a Edge verb.
```
Comment (Cisphyx) [inline synapse/lib/stormlib/model.py:266]: 'an Edge verb'

File: synapse/lib/stormlib/model.py [hunk 21]
```diff
@@ -255,6 +343,12 @@ async def _methEdgeDel(self, verb, key):
         await self.runt.snap.core.setHiveKey(path, kvdict)
 
     async def _methEdgeList(self):
+        '''
+        Get as list of (verb, key-value dictionary) pairs for edges verbs in the current Cortex View.
```
Comment (Cisphyx) [inline synapse/lib/stormlib/model.py:347]: 'Get a list' and 'for Edge verbs''

File: synapse/lib/stormtypes.py [hunk 22]
```diff
@@ -225,10 +369,28 @@ async def _libSvcAdd(self, name, url):
         return await self.runt.snap.core.addStormSvc(sdef)
 
     async def _libSvcDel(self, iden):
+        '''
+        Remove a Storm Service from the Cortex.
+
+        Args:
+            iden (str): The iden of the servie to remove.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:376]: 'servie'

File: synapse/lib/stormlib/macro.py (hunk 19 above)
Comment (vEpiphyte) [reply synapse/lib/stormlib/macro.py]: fixed

File: synapse/lib/stormhttp.py (hunk 18 above)
Comment (vEpiphyte) [reply synapse/lib/stormhttp.py]: fixed

File: synapse/lib/stormlib/model.py (hunk 20 above)
Comment (vEpiphyte) [reply synapse/lib/stormlib/model.py:266]: fixed

File: synapse/lib/stormlib/model.py (hunk 21 above)
Comment (vEpiphyte) [reply synapse/lib/stormlib/model.py:347]: fixed

File: synapse/lib/stormtypes.py (hunk 22 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py]: fixed

File: synapse/lib/stormtypes.py [hunk 23]
```diff
@@ -268,20 +454,73 @@ def addLibFuncs(self):
...
+        return False
+
+    @property
+    def _null(self):
+        '''
+        This constant represents a value of None that can be used in Storm. It is not called like a function, it can
+        be directly used.
+
+        Examples:
+            Creat a dictionary object with a key whose value is null, and call ``$lib.fire()`` with it::
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:506]: 'Create'

File: synapse/lib/stormtypes.py [hunk 24]
```diff
@@ -401,26 +760,86 @@ async def _pprint(self, item, prefix='', clamp=None):
                 await self.runt.printf(fline)
 
     async def _warn(self, mesg, **kwargs):
+        '''
+        Print a warning message to the runtime.
+
+        Args:
+            mesg (str): String to warn.
+
+            **kwargs: **kwargs: Keyword arguments to substitute into the mesg.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:769]: double '**kwargs'

File: synapse/lib/stormtypes.py [hunk 25]
```diff
@@ -468,29 +925,52 @@ async def join(self, sepr, items):
         '''
         Join items into a string using a separator.
 
-        Example:
+        Args:
+            sepr (str): The seperator used to join things with.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:929]: 'sepArator'

File: synapse/lib/stormtypes.py [hunk 26]
```diff
@@ -468,29 +925,52 @@ async def join(self, sepr, items):
...
+            items (list): A list of items to join together.
 
-            $foo = $lib.str.join('.', ('rep', 'vtx', 'tag'))
+        Examples:
+            Join together a list of strings with a dot seperator::
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:933]: 'sepArator'

File: synapse/lib/stormtypes.py [hunk 27]
```diff
@@ -502,34 +982,75 @@ async def _libBytesPut(self, byts):
...
-    async def byNodeData(self, name):
+    async def _byNodeData(self, name):
+        '''
+        Lift nodes which have a given nodedata name set on them.
+
+        Args:
+            name (str): The name to of the nodedata key to lift by.
+
+        Returns:
+            Yields notes to the pipeline. This must be used in conjunction with the ``yield`` keyword.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:1005]: 'notes'

File: synapse/lib/stormtypes.py [hunk 28]
```diff
@@ -805,21 +1443,33 @@ async def _methQueuePut(self, item):
...
             'open': self._methTeleOpen,
-        })
+        }
 
     async def _methTeleOpen(self, url):
         '''
         Open and return a telepath RPC proxy.
+
+        Args:
+            url (str): The Telepath URL to connect too.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:1463]: 'too'

File: synapse/lib/stormtypes.py [hunk 29]
```diff
@@ -1213,39 +1910,71 @@ def __str__(self):
...
+        Returns:
+            str: The name of the current user.
+        '''
         return self.runt.user.name
 
+@registry.registerLib
 class LibGlobals(Lib):
     '''
-    Global persistent Storm variables
+    A Storm Library for interacting global variables which are persistent across the Cortex.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:1946]: 'interacting with'

File: synapse/lib/stormtypes.py (hunk 23 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:506]: fixed

File: synapse/lib/stormtypes.py (hunk 24 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:769]: fixed

File: synapse/lib/stormtypes.py [hunk 30]
```diff
@@ -1645,16 +2476,28 @@ async def _methTextAdd(self, text, **kwargs):
...
-        self.locls.update({
+    def getObjLocals(self):
+        return {
             'tally': self.tally,
-        })
+        }
 
     async def tally(self):
+        '''
+        Get a Tally objectl
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:2493]: 'objectl'

File: synapse/lib/stormtypes.py (hunk 25 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:929]: fixed

File: synapse/lib/stormtypes.py (hunk 26 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:933]: fixed

File: synapse/lib/stormtypes.py (hunk 27 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:1005]: They are very descriptive notes.

Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:1005]: fixed

File: synapse/lib/stormtypes.py (hunk 28 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:1463]: fixed

File: synapse/lib/stormtypes.py (hunk 29 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:1946]: fixed

File: synapse/lib/stormtypes.py [hunk 31]
```diff
@@ -2183,136 +3165,274 @@ def ruleFromText(text):
...
         udef = await self.runt.snap.core.getUserDef(iden)
         if udef is not None:
             return User(self.runt, udef['iden'])
 
     async def _methUsersByName(self, name):
+        '''
+        Get a specific user by name.
+
+        Args:
+            name (str): The name of te user to retrieve.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:3224]: 'te user'

File: synapse/lib/stormtypes.py [hunk 32]
```diff
@@ -2183,136 +3165,274 @@ def ruleFromText(text):
...
     async def _methUsersAdd(self, name, passwd=None, email=None):
+        '''
+        Add a User to the Cortex.
+
+        Args:
+            name (str): The name of the user.
+
+            passwd (str): The users password. This is optional.
+
+            email (str): The user's emaill address. This is optional.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:3242]: 'emaill'

File: synapse/lib/stormtypes.py (hunk 30 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:2493]: fixed

File: synapse/lib/stormtypes.py (hunk 31 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:3224]: fixed

File: synapse/lib/stormtypes.py [hunk 33]
```diff
@@ -2183,136 +3165,274 @@ def ruleFromText(text):
...
         rdef = await self.runt.snap.core.getRoleDef(iden)
         if rdef is not None:
             return Role(self.runt, rdef['iden'])
 
     async def _methRolesByName(self, name):
+        '''
+        Get a specific Role by name.
+
+        Args:
+            name (str): The name of te role to retrieve.
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:3308]: 'te role'

File: synapse/lib/stormtypes.py (hunk 32 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:3242]: fixed

File: synapse/lib/stormtypes.py (hunk 33 above)
Comment (vEpiphyte) [reply synapse/lib/stormtypes.py:3308]: fixed

**Pull request merged**